src/util/buffer.rs

```diff
@@ -0,0 +1,266 @@
+//! A memory buffer with resizing.
+
+use core::nonzero::NonZero;
+use std::num::Int;
+use std::ops::Deref;
+use std::rt::heap;
+use std::mem;
+
+/// A safe wrapper around a heap allocated buffer of Ts, tracking capacity only.
+///
+/// Buffer makes no promises about the actual contents of this memory, that's up
+/// to the user of the structure and can be manipulated using the standard pointer
+/// utilities, accessible through the impl of `Deref<Target=*mut T>` for `Buffer<T>`.
+///
+/// As a result of this hands-off approach, `Buffer`s destructor does not attempt
+/// to drop any of the contained elements; the destructor simply frees the contained
+/// memory.
+///
+/// You can think of `Buffer<T>` as an approximation for `Box<[T]>` where the elements
+/// are not guaranteed to be valid/initialized. It is meant to be used as a building
+/// block for other collections, so they do not have to concern themselves with the
+/// minutiae of allocating, reallocating, and deallocating memory.
+pub struct Buffer<T> {
+    buffer: NonZero<*mut T>,
+    cap: usize
+}
+
+unsafe impl<T: Send> Send for Buffer<T> { }
+unsafe impl<T: Sync> Sync for Buffer<T> { }
+
+impl<T> Buffer<T> {
+    /// Create a new, empty Buffer.
+    ///
+    /// ```
+    /// # use collect::util::buffer::Buffer;
+    ///
+    /// let buffer: Buffer<usize> = Buffer::new();
+    /// assert_eq!(buffer.capacity(), 0);
+    /// ```
+    pub fn new() -> Buffer<T> {
+        Buffer {
+            buffer: unsafe { NonZero::new(heap::EMPTY as *mut T) },
+            cap: 0
+        }
+    }
+
+    /// Create a new buffer with space for cap Ts.
+    ///
+    /// ```
+    /// # use collect::util::buffer::Buffer;
+    ///
+    /// let buffer: Buffer<usize> = Buffer::allocate(128);
+    /// assert_eq!(buffer.capacity(), 128);
+    /// ```
+    pub fn allocate(cap: usize) -> Buffer<T> {
+        if cap == 0 { return Buffer::new() }
+
+        Buffer {
+            buffer: unsafe { allocate(NonZero::new(cap)) },
+            cap: cap
+        }
+    }
+
+    /// Reallocate this buffer to fit a new number of Ts.
+    ///
+    /// ```
+    /// # use collect::util::buffer::Buffer;
+    ///
+    /// let mut buffer: Buffer<usize> = Buffer::allocate(128);
+    /// assert_eq!(buffer.capacity(), 128);
+    ///
+    /// buffer.reallocate(1024);
+    /// assert_eq!(buffer.capacity(), 1024);
+    /// ```
+    pub fn reallocate(&mut self, cap: usize) {
+        *self = if self.cap == 0 || cap == 0 {
```

Comment:
Is this reassigning to self in the reallocate case actually safe with destructors? From tests it appears yes, but I'm not actually confident about this.